### PR TITLE
CLN: Cleanup ops.py

### PR DESCRIPTION
**Pull request by sinhrks**
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`

Cleanup duplicated code paths before fixing some ops related issues.

### CODE REVIEW

File: pandas/core/ops.py [hunk 1]
```diff
@@ -264,7 +263,57 @@ def add_flex_arithmetic_methods(cls, flex_arith_method, radd_func=None,
                 exclude=exclude)
 
 
-class _TimeOp(object):
+class _Op(object):
+
```
Comment (jreback) [inline pandas/core/ops.py]: maybe add a doc-string here? (e.g. its purpose and args passed)

File: pandas/core/ops.py [hunk 2]
```diff
@@ -264,7 +263,57 @@ def add_flex_arithmetic_methods(cls, flex_arith_method, radd_func=None,
...
+        self.lvalues = left
+        self.rvalues = right
+
+    @classmethod
+    def get_op(cls, left, right, name, na_op):
+        """
+        Get op dispatcher, returns _ArithOp or _TimeOp.
+
+        if ``left`` and ``right`` are appropriate for datetime arithmetic with
```
Comment (jreback) [inline pandas/core/ops.py]: ahh I see you did it here, ok maybe just move this before ``__init__`` as then you read it first and get whats going on.

File: pandas/core/ops.py (hunk 1 above)
Comment (sinhrks) [inline pandas/core/ops.py]: Yes, added.

File: pandas/core/ops.py [hunk 3]
```diff
@@ -257,30 +256,87 @@ def add_flex_arithmetic_methods(cls, flex_arith_method,
...
-        # need to make sure that we are aligning the data
         if isinstance(left, ABCSeries) and isinstance(right, ABCSeries):
-            left, right = left.align(right, copy=False)
+            # avoid repaated alignment
+            if not left.index.equals(right.index):
+                left, right = left.align(right, copy=False)
+
+                index, lidx, ridx = left.index.join(right.index, how='outer',
+                                                    return_indexers=True)
+                # convert DatetimeIndex to have UTC
```
Comment (jorisvandenbossche) [inline pandas/core/ops.py]: I don't understand this comment

**Comment by jorisvandenbossche** [comment]
Looks good! Small comment

File: pandas/core/ops.py (hunk 3 above)
Comment (sinhrks) [inline pandas/core/ops.py]: Updated the comment to explain below:

If the ``Series/DataFrame`` have different tz, the indexes are converted to UTC (on 0.18.1). Added tests also.

```
base = pd.DatetimeIndex(['2011-01-01', '2011-01-02', '2011-01-03'], tz='UTC')
idx1 = base.tz_convert('Asia/Tokyo')[:2]
idx2 = base.tz_convert('US/Eastern')[1:]

pd.Series([1, 2], index=idx1) + pd.Series([1, 1], index=idx2)
# 2011-01-01 00:00:00+00:00    NaN
# 2011-01-02 00:00:00+00:00    3.0
# 2011-01-03 00:00:00+00:00    NaN
# dtype: float64
```

Comment (jorisvandenbossche) [inline pandas/core/ops.py]: Sorry, still not fully following. Doesn't the alignment already take care of that?

Using your example from above:

```
In [38]: left = pd.Series([1,2], index=idx1)

In [39]: right = pd.Series([1,2], index=idx2)

In [40]: left, right = left.align(right, copy=False)

In [41]: left.index
Out[41]: DatetimeIndex(['2011-01-01', '2011-01-02', '2011-01-03'], dtype='dateti
me64[ns, UTC]', freq=None)

In [42]: right.index
Out[42]: DatetimeIndex(['2011-01-01', '2011-01-02', '2011-01-03'], dtype='dateti
me64[ns, UTC]', freq=None)
```

Also, the `lidx` and `ridx` does not seem to get used here

Comment (sinhrks) [inline pandas/core/ops.py]: Ah the example was incorrect. Pls try:

```
rng = pd.date_range('1/1/2011', periods=3, freq='H', tz='US/Eastern')
ts = pd.Series(np.random.randn(len(rng)), index=rng)
ts_moscow = ts.tz_convert('Europe/Moscow')\

# UTC
ts + ts_moscow
# 2011-01-01 05:00:00+00:00    1.284372
# 2011-01-01 06:00:00+00:00   -1.729165
# 2011-01-01 07:00:00+00:00   -1.532616
# Freq: H, dtype: float64

# align doesn't
ts.align(ts_moscow)
# 2011-01-01 00:00:00-05:00    0.642186
# 2011-01-01 01:00:00-05:00   -0.864582
# 2011-01-01 02:00:00-05:00   -0.766308
# Freq: H, dtype: float64

# 2011-01-01 08:00:00+03:00    0.642186
# 2011-01-01 09:00:00+03:00   -0.864582
# 2011-01-01 10:00:00+03:00   -0.766308
# Freq: H, dtype: float64
```

**Comment by jreback** [comment]
thxs!

**Pull request closed without merging**